### ISSUE TITLE
Allow templates and droplists for annotations

### ISSUE THREAD

**Issue by o-o00o-o** [otykier/TabularEditor#212]
We are integrating more metadata into our cube that provides users with some useful information as product documentation. We can then expose this via a Power BI report as a single data dictionary. For example

DataOwner
DataSources (i.e. the real data source from the external system) 
SimpleDescription (we can't use description as we autogenerate this from these other metadata columns)
BusinessUse 
etc.

These annotations can be on Tables, Measures and Columns and each of these have their own set of annotations. E.g. DataOwner is only on columns and measures. DataSources is only on Tables and Columns etc.

I thought I would throw it out there that it would be really good to be able to 
1. Have these annotations automatically create on new measures, columns and tables as devs add new ones
2. Provide default template values for each (to provide guidance) 
3. Indicate that some annotations should not be blank (or to ensure that it has been changed from the template value)
4. Be able to validate the contents with a regex (e.g. always from a droplist of discrete items (Marketing, Retail, Finance etc). 

I could write best practice tests and custom actions to achieve some of this (and I may have to) but it would be much more friendly and intuitive for the devs to have this with immediate feedback if they are breaching the rules. A more complete example below

| Annotation Name | Template Value | Scope | RegEx | AllowBlank |
|-|-|-|-|-|
|EachRowIs|Brand data per Date and Lease|Table||No|
|DataSources|European Village data is loaded from system "blah" using the tables dbo.Table1 .Table2 dbo.Table3|Table||No|
|Load Frequency||Table|(Hourly\|Daily\|Adhoc)|No|
|Source Availability|Real Time/Weekly (every Mon)|Table|^(Real Time\|Weekly \(every (Mon\|Tue\|Wed\|Thu\|Fri\|Sat\|Sun))\)$| No|
|DataOwner||Table,Column|^(Marketing\|Retail\|Finance)$|No|
|DataSources|Salesforce.Account.Name|Column|.+\..+\..+|No

I haven't double checked the regexes but hopefully the idea is clear

"Allow blank" would check for blank as well as the current template value to ensure that it has been modified

**Comment by otykier** [comment]
I would use actions and best practice rules to handle this. I realise that editing annotations within the property grid is not ideal for this scenario, but perhaps a solution to #211 that would allow you to import/export them to a TSV file would help. Besides, the roadmap for the Best Practice Analyzer is to have it continuously scan and display rule violations while developing the model, so once this is released, developers would get the immediate feedback needed for something like this.

I would use a script similar to the following, for creating the missing annotations across all measures, columns and tables, and provide the template values. Save the script as a Custom Action at Model scope to make reuse easy. Adjust as necessary:

```
// Ensure annotations exist on tables:
foreach(var t in Model.Tables)
{
    if(t.IsHidden) continue; // Skip hidden tables
    if(t.GetAnnotation("SimpleDescription") == null) t.SetAnnotation("SimpleDescription", ""); // No default - just create a blank annotation
    if(t.GetAnnotation("DataSources") == null) t.SetAnnotation("DataSources", "European Village data is loaded from system "blah" using the tables dbo.Table1 .Table2 dbo.Table3");
    if(t.GetAnnotation("LoadFrequency") == null) t.SetAnnotation("LoadFrequency", "");
    // ... etc. ...
}

// ... repeat above for Model.AllColumns and Model.AllMeasures ...
```

This basically handles point 1 and 2. Then, you'd create a Best Practice Analyzer rule for each type of annotation you have:

```
  {
    "ID": "LOAD_FREQUENCY",
    "Name": "Load Frequency",
    "Category": "Metadata",
    "Description": "Provide a Load Frequency for all tables (Hourly / Daily / Adhoc)",
    "Severity": 1,
    "Scope": "Table",
    "Expression": "GetAnnotation(\"LoadFrequency\") == null or not RegEx.IsMatch(GetAnnotation(\"LoadFrequency\"),\"(Hourly|Daily|Adhoc)\")",
    "FixExpression": null,
    "CompatibilityLevel": 1200
  }
```

![image](https://user-images.githubusercontent.com/8976200/47620457-a2015980-daea-11e8-87c0-0d74132c5e75.png)

For annotations where a template value has been provided, use a rule expression that would catch that, for example:

```
GetAnnotation("DataSources") == null or
GetAnnotation("DataSources") == "European Village data is loaded from system ""blah"" using the tables dbo.Table1 .Table2 dbo.Table3"
```

This handles point 3 and 4.

Also, don't forget you can validate [Best Practice Rules from the command line](https://github.com/otykier/TabularEditor/wiki/Command-line-Options#running-the-best-practice-analyzer) using the -A switch, allowing you to check that annotations have been supplied in your CI/CD pipeline.

**Comment by o-o00o-o** [comment]
Yeah thanks I will do it this way. It was where I was going anyway but there were a couple of limitations. I didn't realise that in the future the BP rules would be realtime checks and this makes a massive difference for this.

The other thing is I guess custom actions that are triggered on UI action. E.g. fire this custom action when a new measure/column/table etc is created/changed etc. Is there anything in the roadmap for this?

I think it would be good to be able to define the default text centrally so both the best practice and the custom action would use a centrally defined text and I guess I could use a model level annotation to achieve that. 

I'll let you know how I get on.

**Comment by o-o00o-o** [comment]
Is there any way of having a Custom Action saved as part of the model so the whole team gets it. I think you mentioned it above, but I can't find any docs or UI that allows doing this?

**Comment by otykier** [comment]
Unfortunately, Custom Actions can not currently be stored within a model. You'd have to distribute the "CustomActions.json" file to be stored in each developers `%AppData%\..\Local\TabularEditor` folder. It makes a lot of sense to be able to store them within a model though, as they are somewhat related to best practice rules, so I'll certainly consider that.

The plug-in framework was originally intended for advanced customisations like some of the stuff you mention. For example, using the framework, it's possible to listen to various events (such as when a measure is added or deleted, etc.), and trigger whatever code you need to run. Unfortunately, it's poorly documented and I'm not even sure if anyone is using it, so I haven't been developing the framework for some time.

**Comment by o-o00o-o** [comment]
I would definitely use the events thing it if it is resurrect-able, but the model stored custom actions would solve more immediate problems for me with this

**Comment by o-o00o-o** [comment]
To try to make this a bit more data driven, I would like to set the regex and the default value as an annotation at model level, then enumerate through them to determine the current values. 

i.e. 
StdAnn_Table_LoadFrequency_Default = "Hourly/Daily/Adhoc (pick one)"
StdAnn_Table_LoadFrequency_Regex = "^(Hourly|Daily|Adhoc)$"

This way I can loop through all annotations in the BP and Custom Action and use the presence of an Annotation called StdAnn_Table for the table BP , StdAnn_Column for the column BP etc.

Is this possible? I can't find a "Annotations" property that allows enumeration only Get and Set methods?

**Comment by otykier** [comment]
This is the definition of the [IAnnotationObject interface](https://github.com/otykier/TabularEditor/blob/f927cf7b97f10c42de6cce4243a6c0c4933ca35e/TOMWrapper/TOMWrapper/Interfaces.cs#L80-L90), which defines the methods available on all objects that support annotations (Tables, Measures, Columns, etc.):

```
    public interface IAnnotationObject: ITabularObject
    {
        string GetAnnotation(int index);
        string GetAnnotation(string name);
        string GetNewAnnotationName();
        void SetAnnotation(int index, string value, bool undoable = true);
        void SetAnnotation(string name, string value, bool undoable = true);
        void RemoveAnnotation(string name, bool undoable = true);
        int GetAnnotationsCount();
        IEnumerable<string> GetAnnotations();
    }
```

The `GetAnnotations()` method returns the names of all annotations on an object. So let's say you defined the two annotations above on the Model object, you could use the following script when creating new tables to apply the defaults to any table that does not yet have them:

```
foreach(var t in Model.Tables)
{
    foreach(var a in Model.GetAnnotation().Where(a => a.StartsWith("StdAnn_Table") && a.EndsWith("Default"))) {
        var annotationName = a.Split("_")[2]; // "LoadFrequency" in this example
        var annotationValue = Model.GetAnnotation(a);
        if(t.GetAnnotation(annotationName) == null) t.SetAnnotation(annotationName, annotationValue);
    }
}
```

For Best Practice Rules, I think it would be best to have one rule per object and annotation type. So for example, the following rule (applied to Tables and Calculated Tables) would capture any that either doesn't contain the "LoadFrequency" annotation, or where the value of the annotation does not conform to the RegEx provided in the model annotation:

```
GetAnnotation("LoadFrequency") == null or
not RegEx.IsMatch(GetAnnotation("LoadFrequency"), Model.GetAnnotation("StdAnn_Table_LoadFrequency_Regex"))
```

**Comment by o-o00o-o** [comment]
Just trying to make this more copy and pasteable by defining some variables but I get the error "No property of field 'var' exists in type 'Table. It may be that I have to put literals in due to the way this script is executed but is there any way of doing this in a BPRule?

``` csharp
var CommonAnnotationName  = "DataOwner"  // Change this for each BPRule

// These are consistent across the model so only change these if you change the convention
var ObjectAnnotationName  = "VR_" + CommonAnnotationName
var RegexAnnotationName   = "DefaultAnnotation_Table_" + CommonAnnotationName + "_Regex"
var DefaultAnnotationName = "DefaultAnnotation_Table_" + CommonAnnotationName + "_Default"

// The Test
GetAnnotation(ObjectAnnotationName) == null or
GetAnnotation(ObjectAnnotationName) == Model.GetAnnotation(DefaultAnnotationName)) or
not RegEx.IsMatch(GetAnnotation(ObjectAnnotationName), Model.GetAnnotation(RegexAnnotationName)) 
```

**Comment by otykier** [comment]
Unfortunately, we can't declare variables in Dynamic LINQ, so you'd have to inline everything. I'm afraid comments are not usable either.

The good news is that you don't have to do a null check, as comparing any non-null value with null will simply yield false.

**Comment by o-o00o-o** [comment]
cool thanks for the tip. back to copy and paste!

**Comment by o-o00o-o** [comment]
Well looks like I did need the null check as this works

``` csharp
GetAnnotation("VR_DataOwner") == null 
or GetAnnotation("VR_DataOwner") == Model.GetAnnotation("DefaultAnnotation_Table_DataOwner_Default")
or not RegEx.IsMatch(GetAnnotation("VR_DataOwner"), Model.GetAnnotation ("DefaultAnnotation_Table_DataOwner_Regex"))
```

but this 
``` csharp
GetAnnotation("VR_DataOwner") == Model.GetAnnotation("DefaultAnnotation_Table_DataOwner_Default")
or not RegEx.IsMatch(GetAnnotation("VR_DataOwner"), Model.GetAnnotation ("DefaultAnnotation_Table_DataOwner_Regex"))
```
gave error "Value cannot be null. Parameter name: input "
I presume with objects that were missing the Annotation on the IsMatch method. Not sure why it wasn't short-circuiting with the logic as you suggested?

**Comment by otykier** [comment]
I'm not exactly sure what you're asking, but if you could express your BPA rule using C#, you could always create a script that loops through various objects, while timing the code to see if that can help you narrow it down:

```csharp
var timingResults = "";
var stopwatch = new System.Diagnostics.Stopwatch();

foreach(var table in Model.Tables)
{
    stopwatch.Reset();
    stopwatch.Start();
    
    // Execute the C# equivalent of your BPA rule here, to measure how long it takes to run:
    var value = table.GetAnnotation("VR_DataOwner");
    var pattern = Model.GetAnnotation ("DefaultAnnotation_Table_DataOwner_Regex");
    if(value != null) System.Text.RegularExpressions.Regex.IsMatch(value, pattern);
    
    stopwatch.Stop();
    timingResults += string.Format("{0}: {1} ms\r\n", table.Name, stopwatch.ElapsedMilliseconds);
}

timingResults.Output();
```

This script will output the following:
![image](https://user-images.githubusercontent.com/8976200/48220401-76565d00-e38f-11e8-8f3c-97d92775fbd6.png)

Is that usable?

**Comment by o-o00o-o** [comment]
Yeah thanks I did that. I guess I was wondering if there was any verbose logging available for the execution time for a BPRule

Another issue is that I have some fields with a fullstop in the name. .How do I get the ResolveObjectPath method to find these measures?

e.g. I am getting the error "Model.Audit Load Errors.No. of Errors in the last 30 days doesn't exist"

**Comment by otykier** [comment]
This is a bug (#216).

**Comment by o-o00o-o** [comment]
is there any way around it? can I escape manually or anything so the parser will find it?

**Comment by otykier** [comment]
Unfortunately, it's a limitation of the [ResolveObjectPath code](https://github.com/otykier/TabularEditor/blob/60f6bde46b5ddd5a8f0cb9f06f1548ce75a4c44d/TabularEditor/Scripting/PropertySerializer.cs#L118), so I don't really see any workarounds, other than creating your own object resolver and using that within the script. But that's quite a bit of work, if you have to support many different types of objects.

I'll try to provide a patched build tonight.

**Comment by o-o00o-o** [comment]
ok that would be great thanks!

**Comment by o-o00o-o** [comment]
As part of the build I need to run some BP rules and run a script. A couple of questions on the command line options:

- Currently the BPrules that I want to run are in the model. Is there any way to have these execute from the commandline? Currently the Command line doc indicates that -A either will use the global rules on the machine (not our machine as is a build server) or a file. If I have to maintain these in a file also, it duplicates the code which I'd rather avoid if possible?

- For the -S Script arg, can you specifiy multiple scripts or is it only one. The script I want to run will set the description attribute from a bunch of annotations, so I guess I need to include it in our -BUILD command. I am guessing that this wont be the only script we might want to run as part of the build and I'd prefer not to have to have it as part of one big script but have a file for each logical script file. 

Thanks!

**Comment by otykier** [comment]
Regarding the ResolveObjectPath functionality, you can now enclose object names with square brackets, in case the name contains a fullstop. For example: `Model.Reseller Sales.[No. of customers]`. The `ResolveObjectPath(...)` method has also been made public, along with the `GetObjectPath(...)` method, to make it easier to use these two. The latter will also properly enclose a name with square brackets, if it contains fullstops.

As to your questions:
- The -A switch should validate any rules in the model, regardless of whether an external file is supplied or not. If not - please let me know.
- Currently, you can only execute one script at a time, but nothing prevents you from executing a script, saving the modified model as a file, then execute additional scripts as subsequent build steps.

**Comment by otykier** [comment]
Gah, I should really stop making changes at 3 am in the morning... Anyway, will make sure both these issues (case-sensitivity and bracket enclosing of object names without fullstop) are resolved with the next release coming later this week.

**Comment by o-o00o-o** [comment]
Hi Daniel, are you including the ability to add custom actions in the model as part of this enhancement for 2.8. Just trying to work out if I need to spend some time working around this now or if this might be available in the next couple of months or so. If the latter I can just document it for this period of time.

One further thought I had on this is that it would be really nice if we could store each Custom Action in an actual .cs file (for GIT Pull Request purposes) then be able to have this serialised with the model somehow? Could you consider this instead of putting it in Custom Annotations as it is a lot easier to edit and code review if it is in its own file. 

Thanks!
